text/0000-natvis.md

```diff
@@ -0,0 +1,428 @@
+
+In rustc, a new `-Z natvis={comma-separated list of .natvis files}` flag
+will be added which instructs the compiler to take the set of .natvis
+files for a given crate and store them in the crate metadata. When running the
+linker, using the `MSVC` toolchain, a `/NATVIS` linker option would be set
+for each .natvis file. This includes .natvis files from all crate dependencies,
+if any exist, as well as the current crate and embed them into the pdb.
+Since the `MSVC` linker is the only one that supports embedding natvis files
+into a pdb, this feature would be specific to the `MSVC` toolchain only.
+
+Cargo would add new syntax in `Cargo.toml` for specifying the list of Natvis
+files to be added to the crate. The new manifest key, `natvis` would be added
+to the `[package]` section. This would be in control of setting the `-Z natvis`
+flag in rustc.
+
+For example:
+
+`Cargo.toml`:
+
+```toml
+cargo-features = ["natvis"]
+
+[package]
+name = "natvis"
+version = "0.1.0"
+edition = "2018"
+natvis = ["src/a.natvis", "src/b.natvis"]
+```
+
+This would generate a call to rustc similar to the following,
+(for simplicity purposes, most of the rustc command line has been removed):
+
+`rustc -Z natvis=path/to/file/a.natvis,path/to/file/b.natvis`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One drawback here is that a lot of types implement the Debug trait which
+already specifies how a type should be viewed when debugging. Implementing
+this RFC would mean a Rust developer would have to manually specify the
+Natvis for a type that may already have implemented the Debug trait which
+would be redundant. Currently, running the Debug trait in the debugger directly
+is not possible and so a manual definition would be required to have a debugger view.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This design provides a simple mechanism for cargo to collect the list of
+.natvis files specified for a given crate and embed them in the resulting
+pdb. It does not need any manual intervention by a Rust developer who is
+consuming such a crate to get the debugging experience to work when it is
+viewed under a debugger that supports the Natvis Framework.
+
+This design does not break any existing usage of cargo or rustc. This new feature
+would be strictly opt-in. The Natvis syntax may not be familiar to many Rust
+developers which may lead to a period of learning the syntax. Since this feature
+would be optional, a consumer of a crate that has natvis definitions for types
+would not need to go through this learning curve. 
+
+Not doing this would keep the existing debugging experience for external Rust crates.
+Most Rust types, outside of the standard library, do not have any debugger views
+defined for them by default which makes them difficult to interpret when viewed
+under a debugger.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many debuggers and languages already address this problem. Some do so in a way
```

Comment:
I think it would be more useful to cover how other languages/compilers use natvis than how natvis itself works. IN particular how the natvis files are specified to the compiler

Reply:
It would be useful to cover how other debuggers (GDB and LLDB in particular) do visualisation and how C/C++ compilers handle things